### PR TITLE
fix(test): AmazonS3-test-v1

### PR DESCRIPTION
**Pull request by franckfotso1**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -41,6 +41,7 @@
     "@koa/cors": "3.1.0",
     "@koa/multer": "3.0.0",
     "@koa/router": "10.0.0",
+    "@types/jest": "^26.0.24",
```
Comment (burgerni10) [inline package.json:44]: Why did you add this package ? It does not seem to be used

File: src/north/amazon/AmazonS3.class.spec.js [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+jest.resetAllMocks()
+})
+
+describe('Amazone S3 north', () => {
+
+  it('should be properly initialized', () => {
+    const AmazonS3Config = config.north.applications[8]
+    const AmazonS3North = new AmazonS3(AmazonS3Config, engine)
+    expect(AmazonS3North.canHandleFiles).toBeTruthy()
+    // expect() tobeContinued
```
Comment (burgerni10) [inline src/north/amazon/AmazonS3.class.spec.js:29]: You can check if methods called from the constructor are called with the expected parameters. You will need to mock the different methods

File: src/north/amazon/AmazonS3.class.spec.js [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    await AmazonS3North.handleFile(filePath)
+
+    // let expectedResult = null
+    let expectedError = null
+
+    try {
+    // expectedResult = await ...
+    } catch (error) {
+      expectedError = error
+    }
```
Comment (burgerni10) [inline src/north/amazon/AmazonS3.class.spec.js:48]: There is no use to rewrite the method handleFile : you must test it, not redo it

File: src/north/amazon/AmazonS3.class.spec.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+    expect(AmazonS3North.params).toBeCalledWith(filePath, bucket, body, key)
```
Comment (burgerni10) [inline src/north/amazon/AmazonS3.class.spec.js:50]: AmazonS3North.params is not a function so you cannot check that it is called with these parameters.

You can instead check for its values. Check on other component for this kind of test

Comment (burgerni10) [reply src/north/amazon/AmazonS3.class.spec.js:50]: You have something similar juste two lines beneath (l 52)

File: src/north/amazon/AmazonS3.class.spec.js [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+    // expect(expectedResult).toBeNull()
+    expect(expectedError).toEqual(ApiHandler.STATUS.COMMUNICATION_ERROR)
```
Comment (burgerni10) [inline src/north/amazon/AmazonS3.class.spec.js:52]: I think the expectedError must be the returned value from handleFile so you must assign it to the return of the method

File: src/north/filewriter/FileWriter.class.spec.js [hunk 6]
```diff
@@ -1,7 +1,11 @@
 const FileWriter = require('./FileWriter.class')
```
Comment (burgerni10) [inline src/north/filewriter/FileWriter.class.spec.js:1]: Please remove this test file from your pr and add a specific pr for it as we discussed

**Review by burgerni10 — CHANGES_REQUESTED** [review]
When you are on ongoing changes, please make it a draft PR so we know it's not for final review

File: package.json (hunk 1 above)
Comment (franckfotso1) [reply package.json:44]: it's for auto-completion on Vs-code

File: src/north/filewriter/FileWriter.class.spec.js (hunk 6 above)
Comment (franckfotso1) [reply src/north/filewriter/FileWriter.class.spec.js:1]: Okay, I will do it , sorry i didn't commit at the right place

File: src/north/amazon/AmazonS3.class.spec.js (hunk 2 above)
Comment (burgerni10) [reply src/north/amazon/AmazonS3.class.spec.js:29]: expect(AmazonS3North.bucket).toBe(AmazonS3Config.AmazonS3.bucket)

**Comment by franckfotso1** [comment]
sur le dernier commit actuel, les tests qui échouent sont liés au fichier FileWriter, étant donné que je ferai une PR spéciale pour le test de ce fichier, j'ai laissé cela comme cela.

File: src/north/amazon/AmazonS3.class.spec.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+  const AmazonS3Config = config.north.applications[8]
+  const AmazonS3North = new AmazonS3(AmazonS3Config, engine)
+
+  it('should be properly initialized', () => {
+    expect(AmazonS3North.bucket).toEqual(AmazonS3Config.AmazonS3.bucket)
+    expect(AmazonS3North.folder).toEqual(AmazonS3Config.AmazonS3.folder)
+    AWS.config.update = jest.fn()
+    const accessKeyId = '' // doubt about what to put here
+    const secretAccessKey = ''
+    expect(AWS.config.update).toHaveBeenCalledWith(accessKeyId, secretAccessKey)
```
Comment (burgerni10) [inline src/north/amazon/AmazonS3.class.spec.js:32]: It should have been called with the id you put in the config. So you don't need to create new variables and you can directly check if the method has been called with AmazonS3Config.AmazonS3(accessKeyId, secretKey)

Be careful, the secret key must be decrypted so it is not the string stored in the config. You must used the return value of the mocked method of the encryption service

File: src/north/amazon/AmazonS3.class.spec.js [hunk 8]
```diff
@@ -0,0 +1,50 @@
+const fs = require('fs')
+const path = require('path')
+const AWS = require('aws-sdk')
```
Comment (burgerni10) [inline src/north/amazon/AmazonS3.class.spec.js:3]: AWS is not mocked, you must mock it

File: src/north/amazon/AmazonS3.class.spec.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+  it('should properly handle file', async () => {
+    const filePath = '/path/to/file/example.file'
+    jest.spyOn(fs, 'createReadStream').mockImplementation(() => ({ key: '' })) // necessary ?
+    const expectedparams = {
+      Bucket: AmazonS3Config.AmazonS3.bucket,
+      Body: AmazonS3Config.AmazonS3.body,
+      Key: `${AmazonS3Config.AmazonS3.folder}/${AmazonS3Config.AmazonS3.getFilenameWithoutTimestamp(filePath)}`,
+      // doubt about the last parameter
+    }
```
Comment (burgerni10) [inline src/north/amazon/AmazonS3.class.spec.js:45]: Bucket comes from the config file, so it is correct, but Body is something else. Try to understand the upload method to understand what is the Body param.
Same problem with the key : there is no method in the config object. Ask yourself what is AmazonS3Config, what's its value when it is set, etc (you can console log its value if it can help you)

File: src/north/amazon/AmazonS3.class.spec.js [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+// Mock engine
+const engine = jest.genMockFromModule('../../engine/Engine.class')
+engine.configService = { getConfig: () => ({ engineConfig: config.engine }) }
+
+beforeEach(() => {
+  jest.resetAllMocks()
+})
+
+describe('Amazone S3 north', () => {
```
Comment (burgerni10) [inline src/north/amazon/AmazonS3.class.spec.js:22]: the function in the describe must be async because you use await inside it

**Review by burgerni10 — CHANGES_REQUESTED** [review]

**Comment by burgerni10** [comment]
I've updated your branch so you will be up to date.
To have the latest changes, update your git tree (git fetch) and from this branch, run the command `git reset --hard origin/Jest-unit-testing`

**Comment by franckfotso1** [comment]
describe('Amazone S3 north',  () => {
  const AmazonS3Config = config.north.applications[8]
  const AmazonS3North = new AmazonS3(AmazonS3Config, engine)  

  it('should be properly initialized', () => {

    expect(AmazonS3North.bucket).toEqual(AmazonS3Config.AmazonS3.bucket)
    expect(AmazonS3North.folder).toEqual(AmazonS3Config.AmazonS3.folder)
    AWS.config.update = jest.fn()
    expect(AWS.config.update).toHaveBeenCalledWith({
      accessKeyId: AmazonS3Config.AmazonS3.authentication.accessKey,
      secretAccessKey: AmazonS3Config.AmazonS3.authentication.secretKey
    })    
    expect(AmazonS3North.s3).toHaveBeenCalledTimes(1)
    expect(AmazonS3North.canHandleFiles).toBeTruthy()    
  })


  it('should properly handle file', async () => {
    const filePath = '/path/to/file/example.file'
    jest.spyOn(fs, 'createReadStream').mockImplementation(() => ({ key: '' })) // necessary ?
    const expectedparams = {
      Bucket: AmazonS3Config.AmazonS3.bucket, 
      Body: AmazonS3Config.AmazonS3.body,
      Key: `${AmazonS3Config.AmazonS3.folder}/${AmazonS3Config.AmazonS3.getFilenameWithoutTimestamp(filePath)}`,
      // doubt about the last parameter 
    }
    let expectedResult = await AmazonS3North.handleFile(filePath)
    expect(AmazonS3North.s3.upload).toHaveBeenCalledWith(expectedparams)
    expect(expectedResult).toEqual(ApiHandler.STATUS.SUCCESS)
  })
})

**Comment by franckfotso1** [comment]
// Mock AWS
jest.mock('aws-sdk',() => {})

**Comment by franckfotso1** [comment]
// Mock EncryptionService
EncryptionService.getInstance = () => ({ decryptText: (password) => password })
